### PR TITLE
[SPARK-19456][SparkR]:Add LinearSVC R API

### PR DESCRIPTION
**Pull request by wangmiao1981**
## What changes were proposed in this pull request?

Linear SVM classifier is newly added into ML and python API has been added. This JIRA is to add R side API.

Marked as WIP, as I am designing unit tests.

## How was this patch tested?

Please review http://spark.apache.org/contributing.html before opening a pull request.

### CODE REVIEW

**Comment by wangmiao1981** [comment]
retest this please

File: R/pkg/R/generics.R [hunk 1]
```diff
@@ -1376,6 +1376,10 @@ setGeneric("spark.kstest", function(data, ...) { standardGeneric("spark.kstest")
 #' @export
 setGeneric("spark.lda", function(data, ...) { standardGeneric("spark.lda") })
 
+#' @rdname spark.linearSvc
+#' @export
+setGeneric("spark.linearSvc", function(data, formula, ...) { standardGeneric("spark.linearSvc") })
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: any name more R familiar that could be more fitting here?

Comment (wangmiao1981) [inline R/pkg/R/generics.R]: svm with linear kernel in glmnet package has the same functionality as this one. But they support other kernels. `linearSvm` is better than `linearSvc` (c denotes classifier)?

Comment (felixcheung) [inline R/pkg/R/generics.R]: maybe `svmLinear`? http://topepo.github.io/caret/available-models.html

Comment (wangmiao1981) [inline R/pkg/R/generics.R]: `svmLinear` looks fine. I will change the files tomorrow. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import org.apache.spark.sql.{DataFrame, Dataset}
+
+private[r] class LinearSVCWrapper private (
+    val pipeline: PipelineModel,
+    val features: Array[String],
+    val labels: Array[String]) extends MLWritable {
+  import LinearSVCWrapper._
+
+  private val svcModel: LinearSVCModel =
+    pipeline.stages(1).asInstanceOf[LinearSVCModel]
```
Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala]: The last state is id_to_index. So I need to use stages(1) to get the fitted model.

File: R/pkg/inst/tests/testthat/test_mllib_classification.R [hunk 3]
```diff
@@ -27,6 +27,44 @@ absoluteSparkPath <- function(x) {
...
+  expect_equal(summary$coefficients, list(-0.1563083, -0.460648, 0.2276626, 1.055085),
+               tolerance = 1e-2)
+  expect_equal(summary$intercept, -0.06004978, tolerance = 1e-2)
+
+  # Test prediction with string label
+  prediction <- predict(model, training)
+  expect_equal(typeof(take(select(prediction, "prediction"), 1)$prediction), "character")
+  expected <- c("versicolor", "versicolor", "versicolor", "virginica",  "virginica",
+                "virginica",  "virginica",  "virginica",  "virginica",  "virginica")
+  expect_equal(sort(as.list(take(select(prediction, "prediction"), 10))[[1]]), expected)
```
Comment (wangmiao1981) [inline R/pkg/inst/tests/testthat/test_mllib_classification.R]: Add `sort` to make it stable.

**Comment by wangmiao1981** [comment]
close to trigger windows test

**Comment by wangmiao1981** [comment]
Open to trigger

**Comment by wangmiao1981** [comment]
@felixcheung I have addressed the comments. cc @yanboliang  @hhbyyh  Thanks!

File: R/pkg/R/mllib_classification.R [hunk 4]
```diff
@@ -39,6 +46,116 @@ setClass("MultilayerPerceptronClassificationModel", representation(jobj = "jobj"
...
+#' summary(savedModel)
+#' }
+#' @note spark.svmLinear since 2.2.0
+setMethod("spark.svmLinear", signature(data = "SparkDataFrame", formula = "formula"),
+          function(data, formula, regParam = 0.0, maxIter = 100, tol = 1E-6, standardization = TRUE,
+                   threshold = 0.5, weightCol = NULL) {
+            formula <- paste(deparse(formula), collapse = "")
+
+            if (is.null(weightCol)) {
+              weightCol <- ""
```
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: is `""` valid as weightCol on the model? would it be better to pass NULL and check for null at the wrapper, if null don't call setWeightCol?

File: R/pkg/R/mllib_classification.R [hunk 5]
```diff
@@ -39,6 +46,116 @@ setClass("MultilayerPerceptronClassificationModel", representation(jobj = "jobj"
...
+#'
+#' # can also read back the saved model and predict
+#' # Note that summary deos not work on loaded model
+#' savedModel <- read.ml(path)
+#' summary(savedModel)
+#' }
+#' @note spark.svmLinear since 2.2.0
+setMethod("spark.svmLinear", signature(data = "SparkDataFrame", formula = "formula"),
+          function(data, formula, regParam = 0.0, maxIter = 100, tol = 1E-6, standardization = TRUE,
+                   threshold = 0.5, weightCol = NULL) {
```
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: threshold defaults to 0.0 on the Scala side?

File: R/pkg/R/mllib_classification.R [hunk 6]
```diff
@@ -39,6 +46,116 @@ setClass("MultilayerPerceptronClassificationModel", representation(jobj = "jobj"
...
+#' @param regParam The regularization parameter.
+#' @param maxIter Maximum iteration number.
+#' @param tol Convergence tolerance of iterations.
+#' @param standardization Whether to standardize the training features before fitting the model. The coefficients
+#'                        of models will be always returned on the original scale, so it will be transparent for
+#'                        users. Note that with/without standardization, the models should be always converged
+#'                        to the same solution when no regularization is applied. Default is TRUE, same as glmnet.
+#' @param threshold The threshold in binary classification, in range [0, 1].
+#' @param weightCol The weight column name.
+#' @param ... additional arguments passed to the method.
```
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: do you think we should add the expert param `aggregationDepth`?

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+  lazy val coefficients: Array[Double] = svcModel.coefficients.toArray
```
Comment (felixcheung) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala]: should this handle coefficients like in SPARK-19395?

File: R/pkg/R/mllib_classification.R [hunk 8]
```diff
@@ -39,6 +46,116 @@ setClass("MultilayerPerceptronClassificationModel", representation(jobj = "jobj"
...
+#' @param data SparkDataFrame for training.
+#' @param formula A symbolic description of the model to be fitted. Currently only a few formula
+#'                operators are supported, including '~', '.', ':', '+', and '-'.
+#' @param regParam The regularization parameter.
+#' @param maxIter Maximum iteration number.
+#' @param tol Convergence tolerance of iterations.
+#' @param standardization Whether to standardize the training features before fitting the model. The coefficients
+#'                        of models will be always returned on the original scale, so it will be transparent for
+#'                        users. Note that with/without standardization, the models should be always converged
+#'                        to the same solution when no regularization is applied. Default is TRUE, same as glmnet.
```
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: is this  "same as glmnet" correct here?

File: R/pkg/R/mllib_classification.R (hunk 4 above)
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: I guess this is the same in spark.logit

File: R/pkg/R/mllib_utils.R [hunk 9]
```diff
@@ -35,7 +35,8 @@
 #' @seealso \link{spark.als}, \link{spark.bisectingKmeans}, \link{spark.gaussianMixture},
 #' @seealso \link{spark.gbt}, \link{spark.glm}, \link{glm}, \link{spark.isoreg},
 #' @seealso \link{spark.kmeans},
-#' @seealso \link{spark.lda}, \link{spark.logit}, \link{spark.mlp}, \link{spark.naiveBayes},
+#' @seealso \link{spark.lda}, \link{spark.logit}, \link{spark.svmLinear},
```
Comment (felixcheung) [inline R/pkg/R/mllib_utils.R]: sort?

File: R/pkg/R/mllib_utils.R [hunk 10]
```diff
@@ -50,7 +51,7 @@ NULL
 #' @seealso \link{spark.als}, \link{spark.bisectingKmeans}, \link{spark.gaussianMixture},
 #' @seealso \link{spark.gbt}, \link{spark.glm}, \link{glm}, \link{spark.isoreg},
 #' @seealso \link{spark.kmeans},
-#' @seealso \link{spark.logit}, \link{spark.mlp}, \link{spark.naiveBayes},
+#' @seealso \link{spark.logit}, \link{spark.svmLinear}, \link{spark.mlp}, \link{spark.naiveBayes},
```
Comment (felixcheung) [inline R/pkg/R/mllib_utils.R]: sort?

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: R/pkg/R/mllib_classification.R (hunk 5 above)
Comment (wangmiao1981) [inline R/pkg/R/mllib_classification.R]: Yes

File: R/pkg/R/mllib_classification.R (hunk 6 above)
Comment (wangmiao1981) [inline R/pkg/R/mllib_classification.R]: In other algorithms, we don't add `aggregationDepth`. Shall I add them in another PR?

File: R/pkg/R/mllib_classification.R (hunk 8 above)
Comment (wangmiao1981) [inline R/pkg/R/mllib_classification.R]: Let me check.

File: R/pkg/R/mllib_classification.R (hunk 4 above)
Comment (wangmiao1981) [inline R/pkg/R/mllib_classification.R]: I think null is better. There are several places like this. Let me double check. Then, I will fix them all in another PR. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala (hunk 7 above)
Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala]: I think this one different. We just want to return it as an Array (list in R).

File: R/pkg/R/mllib_utils.R (hunk 9 above)
Comment (wangmiao1981) [inline R/pkg/R/mllib_utils.R]: OK. I will sort it. Before, it is named as `linearSvc` and I use editor to replace them automatically. Thanks!

File: R/pkg/R/mllib_classification.R (hunk 6 above)
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: I don't think that would hurt. We have expert params in tree models.

File: R/pkg/R/mllib_classification.R (hunk 8 above)
Comment (wangmiao1981) [inline R/pkg/R/mllib_classification.R]: glmnet help message: standardize: Logical flag for x variable standardization, prior to
          fitting the model sequence. The coefficients are always
          returned on the original scale. Default is
          ‘standardize=TRUE’. 
I think they are the same.

Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: but my point is glmnet is linear regression whereas here we are linear svc?
isn't it not a very good reference?

Comment (wangmiao1981) [inline R/pkg/R/mllib_classification.R]: I got your point now. I removed the unnecessary document. Thanks!

File: mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala (hunk 7 above)
Comment (wangmiao1981) [inline mllib/src/main/scala/org/apache/spark/ml/r/LinearSVCWrapper.scala]: I have made the same change as 19395.

File: R/pkg/R/generics.R [hunk 11]
```diff
@@ -1380,6 +1380,10 @@ setGeneric("spark.kstest", function(data, ...) { standardGeneric("spark.kstest")
 #' @export
 setGeneric("spark.lda", function(data, ...) { standardGeneric("spark.lda") })
 
+#' @rdname spark.svmLinear
+#' @export
+setGeneric("spark.svmLinear", function(data, formula, ...) { standardGeneric("spark.svmLinear") })
```
Comment (felixcheung) [inline R/pkg/R/generics.R]: oops, sorry I missed this - we should sort this too

File: R/pkg/R/mllib_classification.R [hunk 12]
```diff
@@ -39,6 +46,131 @@ setClass("MultilayerPerceptronClassificationModel", representation(jobj = "jobj"
...
+#'
+#' # can also read back the saved model and predict
+#' # Note that summary deos not work on loaded model
+#' savedModel <- read.ml(path)
+#' summary(savedModel)
+#' }
+#' @note spark.svmLinear since 2.2.0
+setMethod("spark.svmLinear", signature(data = "SparkDataFrame", formula = "formula"),
+          function(data, formula, regParam = 0.0, maxIter = 100, tol = 1E-6, standardization = TRUE,
+                   threshold = 0.5, weightCol = NULL, aggregationDepth = 2) {
```
Comment (felixcheung) [inline R/pkg/R/mllib_classification.R]: shouldn't we change threashold = 0.0 to match scala as discussed here https://github.com/apache/spark/pull/16800#discussion_r100710324

**Review by felixcheung — CHANGES_REQUESTED** [review]

File: R/pkg/R/mllib_classification.R (hunk 12 above)
Comment (wangmiao1981) [inline R/pkg/R/mllib_classification.R]: Sorry for miss this one. I fix it now.

File: R/pkg/R/generics.R (hunk 11 above)
Comment (wangmiao1981) [inline R/pkg/R/generics.R]: Fixed. Thanks!

**Comment by felixcheung** [comment]
merged to master. could you open a JIRA on programming guide, example, vignettes changes please. Thanks

**Comment by wangmiao1981** [comment]
@felixcheung I will do the example and vignettes today. For the document, I will wait for @hhbyyh to merge his main document first. Thanks!

**Pull request closed without merging**